python/cuml/test/test_metrics.py

```diff
@@ -1103,6 +1105,173 @@ def test_pairwise_distances_output_types(input_type, output_type, use_global):
 
 
+@pytest.mark.parametrize("metric", PAIRWISE_DISTANCE_SPARSE_METRICS.keys())
+@pytest.mark.parametrize("matrix_size", [(3, 8), (800, 10), (2, 400),
+                                         (500, 40)])
+def test_sparse_pairwise_distances(metric: str, matrix_size):
+    # Test the sparse_pairwise_distance helper function.
+    if metric in ["inner_product", "hellinger"]:
+        pytest.skip('Skipping metrics not supported by sklearn')
+    if not has_scipy():
+        pytest.skip('Skipping sparse_pairwise_distances if Scipy is missing')
+    import scipy
+
+    # For fp64, compare at 13 decimals, (2 places less than the ~15 max)
+    compare_precision = 7
+
+    # Compare to sklearn, single input
+    X = scipy.sparse.random(matrix_size[0], matrix_size[1],
+                            dtype=np.float64, random_state=123, density=0.6)
+    S = sparse_pairwise_distances(X, metric=metric)
+    S2 = sklearn_pairwise_distances(X.todense(), metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # Compare to sklearn, double input with same dimensions
+    Y = X
+    S = sparse_pairwise_distances(X, Y, metric=metric)
+    S2 = sklearn_pairwise_distances(X.todense(), Y.todense(), metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # Compare single and double inputs to eachother
+    S = sparse_pairwise_distances(X, metric=metric)
+    S2 = sparse_pairwise_distances(X, Y, metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # Compare to sklearn, with Y dim != X dim
+    Y = scipy.sparse.random(3, matrix_size[1], random_state=123,
+                            density=0.6)
+    S = sparse_pairwise_distances(X, Y, metric=metric)
+    S2 = sklearn_pairwise_distances(X.todense(), Y.todense(), metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # Change precision of one parameter
+    Y = scipy.sparse.random(matrix_size[0], matrix_size[1],
+                            dtype=np.float32, random_state=123, density=0.6)
+    S = sparse_pairwise_distances(X, Y, metric=metric)
+    S2 = sklearn_pairwise_distances(X.todense(), Y.todense(), metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # Change to cupyx with different precision
+    Y = cupyx.scipy.sparse.random(matrix_size[0], matrix_size[1],
+                                  dtype=np.float32,
+                                  random_state=123, density=0.6)
+    S = sparse_pairwise_distances(X, Y, metric=metric)
+    S2 = sklearn_pairwise_distances(X.todense(), Y.todense().get(),
+                                    metric=metric)
+    cp.testing.assert_array_almost_equal(S, S2, decimal=compare_precision)
+
+    # For fp32, compare at 5 decimals, (2 places less than the ~7 max)
+    compare_precision = 4
+
+    # Change precision of both parameters to float
```

Comment:
This test is rather large and it looks like it's testing a lot of stuff. Could the test for changing the precision of one param be a separate test case that works on a very limited number of different sizes (or even a single size)? I think in general we shouldn't need to test scipy's sparse generator against cupy's sparse generator. Could we just use one and copy to host/device as necessary?

Just to test all paths through your code, we probably want to call `pairwise_distances` with the sparse inputs rather than explicitly calling `sparse_pairwise_distances`. For those metrics which sklearn does support in sparse forms, we would preferably compare against the sparse inputs in sklearn, rather than always converting to dense.

Reply:
Now I see that you have a test case below that is a little more like what I was proposing in my previous comment and I'm having trouble seeing the benefit of this test case.

Reply:
This test case is mainly designed to try corner case, I reduced the matrix sizes so that it run more quickly but I think it's important to test the different configurations that are possible.
The test case below `test_sparse_pairwise_distances_sklearn_comparison` is mainly a comparison with larger matrix sizes, that can also run stress tests.
As recommended I modified the sparse generator to only use cupy, with host/device and sparse/dense transfer only when necessary.

Reply:
Sounds good. One little additional nitpick- can we rename this test to something like `test_sparse_corner_cases` for future eyes?